### DISCUSSION TITLE
Command to enable serial console

### DISCUSSION THREAD

**Discussion by Regen379** [hack-technicolor/hack-technicolor#215] [Q&A]
Technicolor TG799vac / VANT-F
Firmware: 17.2.0261-820-RA

https://hack-technicolor.readthedocs.io/en/stable/Hacking/Type2/#final-type-2-steps says to run the following to enable the serial console port

`sed -i -e 's/#//' -e 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`

The second script 's#askconsole:.*\$#askconsole:/bin/ash#' isn't working for me. It doesn't change the 'askconsole' line.

`root@mygateway:~# cat /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

`root@mygateway:~# sed -e 's/#//' -e 's#::askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

I tried some variations to see what worked.

`root@mygateway:~# sed -e 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

`root@mygateway:~# sed 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

`root@mygateway:~# sed -e 's/askconsole:.*\$/askconsole:\/bin\/ash/' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

`root@mygateway:~# sed -e 's/askconsole:\/bin\/login\$/askconsole:\/bin\/ash/' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

`root@mygateway:~# sed -e 's/askconsole:\/bin\/login/askconsole:\/bin\/ash/' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/ash`

`root@mygateway:~# sed -e 's#askconsole:/bin/login#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/ash`

It seems it's happy with using # as a delimiter, but it's not happy with the regular expression .*\$
Try some more variations.

`root@mygateway:~# sed -e 's#askconsole:/bin/login\$#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

`root@mygateway:~# sed -e 's#askconsole:/bin/login$#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/ash`

`root@mygateway:~# sed -e 's#askconsole:.*$#askconsole:/bin/ash#' /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/ash`

`root@mygateway:~# cat /etc/inittab`
`::sysinit:/etc/init.d/rcS S boot`
`::shutdown:/etc/init.d/rcS K shutdown`
`::askconsole:/bin/login`

Further testing (above) shows that it is the escaped end of line '\$' that it doesn't like. It is happy with normal regular expressions.

Does the standard command
`sed -i -e 's/#//' -e 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`
work for anyone else on other models and/or different firmware? I'm seeing the above line on quite a few different websites, so maybe it is just a bug with this model router and/or firmware and/or this version of BusyBox?

`root@mygateway:~# sed --help`
`BusyBox v1.23.2 (2018-06-20 21:43:03 UTC) multi-call binary.`
` `
`Usage: sed [-inrE] [-f FILE]... [-e CMD]... [FILE]...`
`or: sed [-inrE] CMD [FILE]...`
` `
`        -e CMD  Add CMD to sed commands to be executed`
`        -f FILE Add FILE contents to sed commands to be executed`
`        -i[SFX] Edit files in-place (otherwise sends to stdout)`
`                Optionally back files up, appending SFX`
`        -n      Suppress automatic printing of pattern space`
`        -r,-E   Use extended regex syntax`
` `
`If no -e or -f, the first non-option argument is the sed command string.`
`Remaining arguments are input files (stdin if none).`

I'm also wondering if the '$' end of line character is required at all? Wouldn't the '.*' match everything to the end of the line anyway?

UPDATE:
I noticed that when I ran the long multi-line command at
https://hack-technicolor.readthedocs.io/en/stable/Upgrade/#preserving-root-access
the lines

`sed -i 's#/root:.*\$#/root:/bin/ash#' /etc/passwd`
`sed -i -e 's/#//' -e 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`

ended up looking like the following in 
/overlay/$(cat /proc/banktable/booted)/etc/rc.local

`sed -i 's#/root:.*$#/root:/bin/ash#' /etc/passwd`
`sed -i -e 's/#//' -e 's#askconsole:.*$#askconsole:/bin/ash#' /etc/inittab`

i.e. the '\\$' has changed to just '$' on both lines

Similarly, in the screenshot just below "Another 40-50 sec..." at
https://github.com/BoLaMN/tch-exploit
within the section below "RESPONSE" (in white), the green text includes

`sed -i \'s#/root:.*$#/root:/bin/ash#\' /etc/passwd`

This is also missing the '\\' before the '$'.

I'm starting to think that the line

`sed -i -e 's/#//' -e 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`

at https://hack-technicolor.readthedocs.io/en/stable/Hacking/Type2/#final-type-2-steps
is in a format that is suitable to be echoed to a file 
(i.e. as per https://hack-technicolor.readthedocs.io/en/stable/Upgrade/#preserving-root-access)
but is not the right format to be copied into a console/ssh session, and perhaps needs to be corrected?

**Comment by LuKePicci** [comment]
You are probably right. This was copied from the echoed version and I probably forgot to remove that escape. Sorry for the delay but I've been quite busy these days. Thanks for your PR

**Reply by Regen379** [reply]
Hi LuKePicci,
I see someone (you?) has updated https://hack-technicolor.readthedocs.io/en/stable/Hacking/Type2/#final-type-2-steps, removing the extra "\", thanks.

I've just noticed https://www.ilpuntotecnico.com/forum/index.php/topic,77981.msg246548.html#msg246548 has the same command with the extra backslash "\".

`sed -i -e 's/#//' -e 's#askconsole:.*\$#askconsole:/bin/ash#' /etc/inittab`

Should this be changed too (remove the extra backslash)?

`sed -i -e 's/#//' -e 's#askconsole:.*$#askconsole:/bin/ash#' /etc/inittab`
